src/app/machines/views/MachineList/MachineList.test.tsx

```diff
@@ -705,4 +714,106 @@ describe("MachineList", () => {
     expect(fetches).toHaveLength(2);
     expect(fetches[fetches.length - 1].payload.params.page_number).toBe(2);
   });
+
+  it("shows a warning notification if not all controllers are configured with Vault", async () => {
+    const controllers = [
+      controllerFactory({
+        system_id: "abc123",
+        vault_configured: true,
+        node_type: NodeType.REGION_CONTROLLER,
+      }),
+      controllerFactory({
+        system_id: "def456",
+        vault_configured: false,
+        node_type: NodeType.REGION_AND_RACK_CONTROLLER,
+      }),
+    ];
+    state.controller.items = controllers;
+
+    renderWithBrowserRouter(
+      <MachineList searchFilter="" setSearchFilter={jest.fn()} />,
+      { wrapperProps: { state } }
+    );
+
+    expect(screen.getByTestId("vault-notification")).toHaveTextContent(
+      "Configure 1 other controller with Vault to complete this operation."
+    );
+  });
+
+  it("shows a warning notification if  all controllers are configured with Vault but secrets are not migrated", async () => {
+    const controllers = [
+      controllerFactory({
+        system_id: "abc123",
+        vault_configured: true,
+        node_type: NodeType.REGION_CONTROLLER,
+      }),
+      controllerFactory({
+        system_id: "def456",
+        vault_configured: true,
+        node_type: NodeType.REGION_AND_RACK_CONTROLLER,
+      }),
+    ];
+    state.controller.items = controllers;
+
+    renderWithBrowserRouter(
+      <MachineList searchFilter="" setSearchFilter={jest.fn()} />,
+      { wrapperProps: { state } }
+    );
+
+    expect(screen.getByTestId("vault-notification")).toHaveTextContent(
+      "Migrate your secrets to Vault to complete this operation."
+    );
+  });
```

Comment:
```suggestion
    expect(screen.getByText(
      "Migrate your secrets to Vault to complete this operation."
    ).toBeInTheDocument();
  });
```

Reply:
This doesn't work since the notification text content is broken up by multiple elements, as it contains links

Reply:
This worked for the "migrate" text, but not for the "configure x other controllers" text; however, this worked:

```tsx
expect(
  screen.getByText(/Incomplete Vault integration/)
).toBeInTheDocument();
expect(screen.getByText(/Configure 1 other/)).toBeInTheDocument();
expect(
  screen.getByRole("link", { name: "controller" })
).toBeInTheDocument();
expect(screen.getByText(/with Vault to complete this operation./));
```

Reply:
You could use a custom matcher function, like below. I agree this is probably too complex and feels a little hacky.

```tsx
      screen.getByText((content, element) => {
        return (
          content !== "" &&
          element?.textContent?.includes(
            "Configure 1 other controller with Vault to complete this operation."
          )
        );
      })
```

The real solution would be to add proper semantics to notification so that you can query their content by a title (something to solve in vanilla/react-components). I'll add a ticket to look into this. Happy to merge this as is for now.

Reply:
https://github.com/canonical/maas-ui/issues/4516